dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -136,23 +148,31 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
 
     winrt::Windows::Foundation::IInspectable PushNotificationManager::Deserialize(winrt::Windows::Foundation::Uri const& uri)
     {
-        // Verify if the uri contains a background notification payload.
-        // Otherwise, we expect to process the notification in a background task.
-        for (auto const& pair : uri.QueryParsed())
+        // All packaged processes are triggered through COM via Long Running Process or the Background Infra OS component
+        if (AppModel::Identity::IsPackagedProcess())
         {
-            if (pair.Name() == L"payload")
+            THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_TIMEOUT), !m_waitHandleForArgs.wait(c_receiveArgsTimeoutInMSec));
+            auto lock{ m_lock.lock_shared() };
+            THROW_HR_IF(E_UNEXPECTED, !m_backgroundTaskArgs);
+            return m_backgroundTaskArgs;
+        }
+        else // The process was launched via ShellExecute and we need to parse the uri (Only unpackaged)
+        {
+            for (auto const& pair : uri.QueryParsed())
             {
-                // Convert escaped components to its normal content
-                // from the conversion done in the LRP (see NotificationListener.cpp)
-                std::wstring payloadAsEscapedWstring{ pair.Value() };
-                std::wstring payloadAsWstring{ winrt::Windows::Foundation::Uri::UnescapeComponent(payloadAsEscapedWstring) };
-                return winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(payloadAsWstring);
+                if (pair.Name() == L"payload")
+                {
+                    // Convert escaped components to its normal content
+                    // from the conversion done in the LRP (see NotificationListener.cpp)
+                    std::wstring payloadAsEscapedWstring{ pair.Value() };
+                    std::wstring payloadAsWstring{ winrt::Windows::Foundation::Uri::UnescapeComponent(payloadAsEscapedWstring) };
+                    return winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationReceivedEventArgs>(payloadAsWstring);
+                }
             }
-        }
 
-        THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_TIMEOUT), !m_waitHandleForArgs.wait(c_receiveArgsTimeoutInMSec));
-        THROW_HR_IF(E_UNEXPECTED, !m_backgroundTaskArgs);
-        return m_backgroundTaskArgs;
+            THROW_HR_MSG(E_UNEXPECTED, "Could not serialize payload from command line Uri!");
+            return nullptr;
```

Comment:
Is there a way we can restructure this so we don't have unreachable code? 

Suggestion: Set a result variable PushNotificationEventArgs at the front of the if/else and assign it as needed. In the for loop, break on success and if !found throw E_UNEXPECTED.

Reply:
Cleaned this up.